liborbum/src/Controller/Ee/Vpu/Vu/Interpreter/CVuInterpreter.cpp

```diff
@@ -1,53 +1,246 @@
+#include <boost/format.hpp>
+
 #include "Controller/Ee/Vpu/Vu/Interpreter/CVuInterpreter.hpp"
 
 #include "Core.hpp"
 #include "Resources/RResources.hpp"
 
 CVuInterpreter::CVuInterpreter(Core* core) :
-    CController(core)
+    CVu(core)
 {
 }
 
-void CVuInterpreter::handle_event(const ControllerEvent& event)
+int CVuInterpreter::time_step(const int ticks_available)
 {
-    switch (event.type)
-    {
-    case ControllerEvent::Type::Time:
-    {
-        int ticks_remaining = time_to_ticks(event.data.time_us);
-        while (ticks_remaining > 0)
-            ticks_remaining -= time_step(ticks_remaining);
-        break;
-    }
-    default:
+#if defined(BUILD_DEBUG)
+    DEBUG_LOOP_COUNTER++;
+#endif
+
+    RResources& r = core->get_resources();
+
+    int cycles_consumed[2] = { 1, 1 };
+
+    // The priority of register writing is: COP2 Transfer > Upper Inst > Lower Inst
+
+    for (auto* unit : r.ee.vpu.vu.units)
     {
-        throw std::runtime_error("CVuInterpreter event handler not implemented - please fix!");
-    }
+        // Move on if the unit is not running and the delay slot is empty
+        if (unit->operation_state != VuOperationState::Run && !unit->bdelay.is_branch_pending())
+        {
+            continue;
+        }
+
+        // PC & Instructions stuff...
+        const uword pc = unit->pc.read_uword() & 0x0FFF;
+        const udword raw_inst = r.ee.bus.read_uword(BusContext::Vu, 0x11000000 + 0x8000 * unit->core_id + pc);
```

Comment:
This should be accessible from the vu unit's bus I think...
`core->bus.read_uword(pc);`

Reply:
Tried that before but interestingly... micro mem is not mapped to the VU bus